src/hypofuzz/dashboard.py

```diff
@@ -50,89 +52,249 @@
+    @staticmethod
+    def _assert_reports_ordered(
+        reports: list[ReportWithDiff], attributes: list[str]
+    ) -> None:
+        for attribute in attributes:
             assert all(
                 getattr(r1, attribute) <= getattr(r2, attribute)
-                for r1, r2 in zip(self.reports, self.reports[1:])
-            ), (attribute, [getattr(r, attribute) for r in self.reports])
-
-        # we track a separate attibute for the total count for efficiency, but
-        # they should be equal.
-        assert self.status_counts == sum(
-            self.reports_offsets.status_counts.values(),
-            start=StatusCounts(dict.fromkeys(Status, 0)),
+                for r1, r2 in zip(reports, reports[1:])
+            ), (attribute, [getattr(r, attribute) for r in reports])
+
+    def _check_invariants(self) -> None:
+        # this entire function is pretty expensive, move from run-time to test-time
+        # once we're more confident
+        self._assert_reports_ordered(self.linear_reports, ["timestamp_monotonic"])
+
+        linear_status_counts = self.linear_status_counts(since=None)
+        assert all(
+            v1 <= v2 for v1, v2 in zip(linear_status_counts, linear_status_counts[1:])
+        ), linear_status_counts
+
+        linear_elapsed_time = self.linear_elapsed_time(since=None)
+        assert all(
+            v1 <= v2 for v1, v2 in zip(linear_elapsed_time, linear_elapsed_time[1:])
+        ), linear_elapsed_time
+        assert (
+            len(linear_elapsed_time)
+            == len(linear_status_counts)
+            == len(self.linear_reports)
         )
+
+        for worker_uuid, reports in self.reports_by_worker.items():
+            assert {r.nodeid for r in reports} == {self.nodeid}
+            assert {r.database_key for r in reports} == {self.database_key}
+            assert {r.worker.uuid for r in reports} == {worker_uuid}
+            self._assert_reports_ordered(self.linear_reports, ["timestamp_monotonic"])
+
         # this is not always true due to floating point error accumulation.
-        # assert self.elapsed_time == sum(self.reports_offsets.elapsed_time.values())
+        # total_elapsed_time = 0.0
+        # for reports in self.reports_by_worker.values():
+        #     total_elapsed_time += reports[-1].elapsed_time
+        # assert self.elapsed_time == total_elapsed_time
 
     def add_report(self, report: Report) -> None:
-        status_counts = self.reports_offsets.status_counts
-        elapsed_time = self.reports_offsets.elapsed_time
-        counts_diff = report.status_counts - status_counts.setdefault(
-            report.worker.uuid, StatusCounts(dict.fromkeys(Status, 0))
+        # we use last_report to compute timestamp_monotonic, and last_report_worker
+        # to compute status_count_diff and elapsed_time_diff
+        last_report = self.linear_reports[-1] if self.linear_reports else None
+        last_report_worker = (
```

Comment:
Conceptually, we should use `last_report_worker` for all attributes including `timestamp_monotonic`.  That does require the ability to insert not just append, so we can defer it for now.

Reply:
It only breaks monotonicity if you append to `linear_reports` instead of bisect-inserting; which is a followup for later

Reply:
hmm, fair, I guess you don't lose anything if you simply compute relative to the worker and inserting at the right location